### PR TITLE
[air/tune] Internal resource management 2 - Ray Tune to use new Ray AIR resource manager

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

Includes/depends on #30777

TLDR: This PR refactors Ray Tune's resource management to use a central AIR resource management package instead of the tightly coupled PlacementGroupManager.

Ray Tune's resource management currently uses a tightly coupled placement group manager. This leads to a number of shortcomings:
- The tight coupling on the manager side (e.g. PG manager keeps track of trials) prevents re-usability
- The tight coupling on the trial executor side prevents using different resource management strategies (e.g. shared or budget-based)
- It's hard to test independently. Tests for the resource management require a simulated tune setup.

To improve stability, extensibility, and maintainability, this PR moves the resource management logic into a central `ray.air.execution.resources` subpackage. The resource management has a simple API that works with `ResourceRequest`s and `AllocatedResources` to manage requested and assigned resources, respectively. The actual resource management can then be anything - per default it is a placement group based manager, but this PR also introduces a PoC budget-based manager that can be plugged in.

The PR does not substantially change existing tests, so we can be certain that the new resource model is a fully compatible replacement for the old placement group manager.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/air/execution/resources/virtual.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    bundles: List[Dict[str, float]]
+
+    def annotate_remote_objects(
+        self, objects
+    ) -> List[Union[ray.ObjectRef, ray.actor.ActorHandle]]:
+        return objects
+
+
+@DeveloperAPI
+class VirtualResourceManager(FixedResourceManager):
```
Comment (xwjiang2010) [inline python/ray/air/execution/resources/virtual.py:23]: what is this for?

File: python/ray/air/execution/resources/resource_manager.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import ray
+from ray.air.execution.resources.request import (
+    ResourceRequest,
+    AllocatedResource,
+)
+from ray.util.annotations import DeveloperAPI
+
+
+@DeveloperAPI
+class ResourceManager(abc.ABC):
```
Comment (xwjiang2010) [inline python/ray/air/execution/resources/resource_manager.py:14]: IMO, this interface is very tailored towards the limitations of placement group interface and not generic enough. Should `get_source_futures` and `update_state` etc be moved down to its PlacementGroupResourceManager?

File: python/ray/tune/execution/ray_trial_executor.py [hunk 3]
```diff
@@ -235,43 +234,50 @@ def __init__(
...
+        # For printing used resources
         self._resource_updater = _ResourceUpdater(refresh_period)
 
-        self._has_cleaned_up_pgs = False
+        # Actor re-use
         self._reuse_actors = reuse_actors
-        # The maxlen will be updated when `setup(max_pending_trials)` is called
-        self._cached_actor_pg = deque(maxlen=1)
-        self._pg_manager = _PlacementGroupManager(prefix=_get_tune_pg_prefix())
+        self._max_staged_actors = 1
```
Comment (xwjiang2010) [inline python/ray/tune/execution/ray_trial_executor.py:242]: this is used for more than reusing actors right?

File: python/ray/tune/execution/ray_trial_executor.py [hunk 4]
```diff
@@ -310,43 +316,77 @@ def _stage_and_update_status(self, trials: Iterable[Trial]):
...
+
             if trial in self._staged_trials:
                 continue
-            if self._pg_manager.trial_in_use(trial):
-                continue
 
-            if not self._pg_manager.stage_trial_pg(trial):
-                # Break if we reached the limit of pending placement groups.
-                break
+            resource_request = trial.placement_group_factory
```
Comment (xwjiang2010) [inline python/ray/tune/execution/ray_trial_executor.py:329]: internally should we unify on `resource_request`?

File: python/ray/tune/execution/ray_trial_executor.py [hunk 5]
```diff
@@ -235,43 +234,50 @@ def __init__(
...
-        self._has_cleaned_up_pgs = False
+        # Actor re-use
         self._reuse_actors = reuse_actors
-        # The maxlen will be updated when `setup(max_pending_trials)` is called
-        self._cached_actor_pg = deque(maxlen=1)
-        self._pg_manager = _PlacementGroupManager(prefix=_get_tune_pg_prefix())
+        self._max_staged_actors = 1
+        self._cached_resources_to_actor = defaultdict(list)
+        self._last_cached_actor_cleanup = float("-inf")
+        # Protect these actors from the first cleanup
```
Comment (xwjiang2010) [inline python/ray/tune/execution/ray_trial_executor.py:245]: What is the problem it's trying to solve?

File: python/ray/tune/execution/ray_trial_executor.py [hunk 6]
```diff
@@ -732,37 +817,75 @@ def has_resources_for_trial(self, trial: Trial) -> bool:
...
 
-        self._pg_manager.cleanup()
+    def _count_staged_resources(self):
+        counter = Counter()
+        for trial in self._staged_trials:
+            resource_request = trial.placement_group_factory
+            counter[resource_request] += 1
+        return counter
+
+    def _cleanup_cached_actors(self, force_all: bool = False):
```
Comment (xwjiang2010) [inline python/ray/tune/execution/ray_trial_executor.py:859]: we should document why do we need this?
Or why is this "cron job" necessary for keeping consistency?

**Comment by xwjiang2010** [comment]
For RayTrialExecutor class, we should document on a high level the "equality" and "inequality" relationship of variables.
case 1: w/o reuse actors
- the number of "staged" trials should be larger than 1 but lower than a certain threshold. This is designed so to trigger autoscaling behavior of the cluster.
- ...
case 2: w reuse actors
- the number of cached reused actor + the number of requested placement group should be less than the number of remaining trials to run, and this is the purpose of yyy method.
- ...

Would that be helpful for understanding?

File: python/ray/air/execution/resources/resource_manager.py (hunk 2 above)
Comment (krfricke) [reply python/ray/air/execution/resources/resource_manager.py:14]: Both are needed because Core doesn't provide any events. We need to keep the interface though to be able to plug in different resource managers.

The default for the resource futures is to just return an empty list, and update state is a no-op. I'll add documentation to point out the contract with these methods.

Just for context, calling these methods is not *required*, it's just a convenience to be able to await different futures (e.g. training and resource futures) like we do in ray tune right now. As an alternative users can just do

```
while not resource_manager.has_resources_ready(request):
    time.sleep(0.1)
```

(but we don't want to do this in Ray Tune)

File: python/ray/air/execution/resources/resource_manager.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+        """Returns True if resources for the given request are available"""
+        raise NotImplementedError
+
+    def acquire_resources(
+        self, resource_request: ResourceRequest
+    ) -> Optional[AllocatedResource]:
+        """Acquire resources. Returns None if resources are not available."""
+        raise NotImplementedError
+
+    def free_resources(self, allocated_resources: AllocatedResource):
```
[the remark this answers is not in the document]
Comment (krfricke) [reply python/ray/air/execution/resources/resource_manager.py:37]: One acts on a resource request and one on allocated resources. I can expand the docstrings a bit

File: python/ray/air/execution/resources/placement_group.py [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+        self._request_to_ready_pgs: Dict[
+            ResourceRequest, Set[PlacementGroup]
+        ] = defaultdict(set)
+
+        self._staging_future_to_pg: Dict[ray.ObjectRef, PlacementGroup] = dict()
+        self._pg_to_staging_future: Dict[PlacementGroup, ray.ObjectRef] = dict()
+        self._acquired_pgs: Set[PlacementGroup] = set()
+
+        self._update_interval = update_interval
+        self._last_update = time.monotonic() - self._update_interval - 1
```
[the remark this answers is not in the document]
Comment (krfricke) [reply python/ray/air/execution/resources/placement_group.py:77]: Because `time.monotonic()` does not have any guarantees around the value it emits, only that two sequential calls are monotonically increasing:

```
The reference point of the returned value is undefined, so that only the difference between the results of two calls is valid.
```

I guess `float("-inf")` would be an alternative

File: python/ray/tune/execution/ray_trial_executor.py [hunk 9]
```diff
@@ -119,8 +117,6 @@ def _post_stop_cleanup(future, pg, timeout: Optional[float] = None):
                 f"An exception occurred when trying to stop the Ray actor:"
                 f"{traceback.format_exc()}"
             )
-    finally:
-        remove_placement_group(pg)
```
[the remark this answers is not in the document]
Comment (krfricke) [reply python/ray/tune/execution/ray_trial_executor.py:123]: With the new abstraction, the whole lifecycle management of resources should be contained in the resource manager.

This will not leak as we free the resources before we call `_post_stop_cleanup`. Actually, let me double check that `RayTrialExecutor._do_force_trial_cleanup` does not leak anything here, but it should be a simple call to `free_resources`

File: python/ray/tune/execution/ray_trial_executor.py [hunk 10]
```diff
@@ -732,37 +817,75 @@ def has_resources_for_trial(self, trial: Trial) -> bool:
...
         return (
             trial in self._staged_trials
-            or (
-                len(self._cached_actor_pg) > 0
-                and (self._pg_manager.has_cached_pg(trial.placement_group_factory))
-            )
-            or self._pg_manager.can_stage()
-            or self._pg_manager.has_ready(trial, update=True)
-            or self._pg_manager.has_staging(trial)
+            or self._cached_resources_to_actor[resource_request]
```
[the remark this answers is not in the document]
Comment (krfricke) [reply python/ray/tune/execution/ray_trial_executor.py:824]: It's a `defaultdict` so will always find a key

File: python/ray/tune/execution/ray_trial_executor.py (hunk 3 above)
Comment (krfricke) [reply python/ray/tune/execution/ray_trial_executor.py:242]: That's right, it generally controls how much autoscaling we trigger (and how many configuration we generate - e.g. we support creating infinite trials)

File: python/ray/tune/execution/ray_trial_executor.py (hunk 4 above)
Comment (krfricke) [reply python/ray/tune/execution/ray_trial_executor.py:329]: Yes, moving pgf to resource request is part of a separate follow-up PR (https://github.com/ray-project/ray/pull/30028, WIP and not updated with the recent changes)

File: python/ray/tune/execution/ray_trial_executor.py (hunk 5 above)
Comment (krfricke) [reply python/ray/tune/execution/ray_trial_executor.py:245]: At the moment of caching, we haven't created a new trial, yet. Thus we don't know if we actually need to cache this actor or not. We only know this on the next iteration of `step()`.

I can add a more thorough comment here

File: python/ray/tune/execution/ray_trial_executor.py (hunk 6 above)
Comment (krfricke) [reply python/ray/tune/execution/ray_trial_executor.py:859]: I think there's two things here

1) (Gracefully) clean up cached actors (e.g. on interrupt) - we need this anyway
2) Scale down cached actors according to resource requirements

The main issue is when a trial completes and we decide to cache or not cache the actor, we don't know the next trial requirements, yet. We only know this after the next step() iteration when we created (or not created) a new trial.

Thus we eagerly cache actors and then check on the next step if we actually still need the actor. Note that this is not a full reconciliation (like PGs before) - we only _remove_ cached actors, never add.

File: python/ray/air/execution/resources/virtual.py (hunk 1 above)
Comment (krfricke) [reply python/ray/air/execution/resources/virtual.py:23]: It reserves resources but does not bind the resources to the remote objects. If I recall correctly it was needed for some use cases of the prototype (I think for some nesting cases)

File: python/ray/tune/execution/ray_trial_executor.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (krfricke) [reply python/ray/tune/execution/ray_trial_executor.py:123]: Turns out the force cleanup logic was missing the PG removal, so there was a leak. Thanks for spotting this, I've updated the PR. We free the resources in `_do_force_trial_cleanup` - this still means lifecycle management is contained in the resource manager.

File: python/ray/air/execution/resources/resource_manager.py (hunk 2 above)
Comment (krfricke) [reply python/ray/air/execution/resources/resource_manager.py:14]: As discussed offline, let's leave it with the dummy interface. If we ever get to the point where we have an event-based manager we can just remove this interface.

**Comment by xwjiang2010** [comment]
Checking my understanding here on the follow up PRs:
- Combine runner and executor.
- have ActorManager that takes care of tune specific actor logic like reusing actors

Also re-raising this point

> For RayTrialExecutor class, we should document on a high level the "equality" and "inequality" relationship of variables.
> case 1: w/o reuse actors
> 
> the number of "staged" trials should be larger than 1 but lower than a certain threshold. This is designed so to trigger autoscaling behavior of the cluster.
> ...
> case 2: w/ reuse actors
> 
> the number of cached reused actor + the number of requested placement group should be less than the number of remaining trials to run, and this is the purpose of yyy method.
> ...
> Would that be helpful for understanding?

Will take another look after 1 is merged.

**Comment by krfricke** [comment]
Yeah that's about right, but I think it's easier to do this the other way around:

1. Introduce actor+future manager, replacing much of the low-level functionality in the trial executor
2. Consolidate trial runner + executor, preparing controller interface
3. Use general controller interface

For the other point, I'm not sure if I follow. 

1. The maximum number of staged actors is independent of actor reuse. But if we re-use an actor, we can unstage another equivalent actor when we cache. This is described in `_maybe_use_cached_actor` (I've expanded the comment a bit just now)
2. We only want to cache actors that we actually want to reuse later. I think this behavior is well described in `_cleanup_cached_actors`

File: python/ray/tune/execution/ray_trial_executor.py [hunk 11]
```diff
@@ -497,12 +533,90 @@ def _start_trial(self, trial: Trial) -> bool:
...
+        # Remove staged trial with same resource requirements
+        candidate_trial = None
+        for staged_trial in self._staged_trials:
+            staged_resources = staged_trial.placement_group_factory
+            if staged_resources == resource_request:
+                candidate_trial = staged_trial
+                break
+
+        if candidate_trial:
+            self._staged_trials.remove(candidate_trial)
```
Comment (xwjiang2010) [inline python/ray/tune/execution/ray_trial_executor.py:563]: why do we keep track of `_staged_trials` in the first place, since everything is around resources? And seems we remove random trials that have the same resources request.

File: python/ray/tune/execution/ray_trial_executor.py [hunk 12]
```diff
@@ -310,43 +316,79 @@ def _stage_and_update_status(self, trials: Iterable[Trial]):
...
-            self._has_cleaned_up_pgs = True
-
         for trial in trials:
+            if len(self._staged_trials) >= self._max_staged_actors:
+                break
+
             if trial.status not in (Trial.PENDING, Trial.PAUSED):
                 continue
+
             if trial in self._staged_trials:
```
Comment (xwjiang2010) [inline python/ray/tune/execution/ray_trial_executor.py:326]: seems like what is needed is just to keep a number of staged trials for each resource type?

File: python/ray/tune/execution/ray_trial_executor.py (hunk 11 above)
Comment (krfricke) [reply python/ray/tune/execution/ray_trial_executor.py:563]: This is existing logic. 

It's the callers responsibility to request resources and cancel resource requests. We only want to request each trial resources once, so we need to keep track for which trials we've already requested resources.

When we reuse actors we cancel the resource request from the trial that will reuse the cached actor. If that trial has not been staged (e.g. because max staging was reached and scheduler decided to start this trial), we can just remove another trial with the same resource requests instead.

File: python/ray/tune/execution/ray_trial_executor.py (hunk 12 above)
Comment (krfricke) [reply python/ray/tune/execution/ray_trial_executor.py:326]: Hm yeah that should work. I can give it a try. Per default I think it's ok to keep it as is as it's existing logic

File: python/ray/tune/execution/ray_trial_executor.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (xwjiang2010) [reply python/ray/tune/execution/ray_trial_executor.py:123]: We have various places calling `post_stop_cleanup`. What is the reason of calling ResourceManager.free_resources() higher in the calling stack instead of here?

File: python/ray/tune/execution/ray_trial_executor.py (hunk 11 above)
Comment (xwjiang2010) [reply python/ray/tune/execution/ray_trial_executor.py:563]: yeah understand the code has always been here. 

However, my understanding is that we shouldn't keep *which trials* are staged and has a bunch of comments explaining the concept of swapping them. 
Could we establish the minimal invariants (sort of tied to my one of my comments about high level invariants that this executor class is operating under). For example, we could state that at any given point, sum(staged_trial_number for any given resource request type) <= `executor._max_staged_actors` (is this correct?) Any many other constraints like this...

I feel like establishing this kind of invariants and state them clearly at the top of the class could greatly reduce the complexity of understanding how this class operates. Let me know if this makes sense.

File: python/ray/tune/execution/ray_trial_executor.py [hunk 13]
```diff
@@ -235,43 +234,50 @@ def __init__(
...
+        # For printing used resources
         self._resource_updater = _ResourceUpdater(refresh_period)
 
-        self._has_cleaned_up_pgs = False
+        # Actor re-use
         self._reuse_actors = reuse_actors
-        # The maxlen will be updated when `setup(max_pending_trials)` is called
-        self._cached_actor_pg = deque(maxlen=1)
-        self._pg_manager = _PlacementGroupManager(prefix=_get_tune_pg_prefix())
+        self._cached_resources_to_actor = defaultdict(list)
```
Comment (xwjiang2010) [inline python/ray/tune/execution/ray_trial_executor.py:242]: Is this actually `_resources_to_cached_actor`?

File: python/ray/tune/execution/ray_trial_executor.py [hunk 14]
```diff
@@ -732,37 +820,106 @@ def has_resources_for_trial(self, trial: Trial) -> bool:
...
+
+        for resource_request, actors in self._cached_resources_to_actor.items():
+            re_add = []
+
+            while len(actors) > staged_resources.get(resource_request, 0) or (
+                force_all and len(actors)
+            ):
+                actor, allocated_resources = actors.pop()
+                if actor in self._newly_cached_actors and not force_all:
+                    self._newly_cached_actors.remove(actor)
```
Comment (xwjiang2010) [inline python/ray/tune/execution/ray_trial_executor.py:907]: hmmm, I am trying to think what is the invariant here. 
The current logic (which is based on if the actor is newly cached, and this only spares them from being cleaned up for one tune loop or whatever the timer says, 0.5s for now), doesn't seem to work well heterogeneous resource types, does it?
It seems, mostly likely, there will only be one more trial added to be pending in the next tune loop. This means only one cached actor of such resource type is meaningfully reused. All the others will be just discarded. Is that efficient?
Should we just do everything time based? We can clearly say that there is a grace period of xxx seconds for each resource type to allow for exact one newly cached actor to be spared from being cleaned up. 
i.e. 
len(cached_actor) <= num_staged + 1, for each resource type.

File: python/ray/tune/execution/ray_trial_executor.py [hunk 15]
```diff
@@ -497,12 +533,90 @@ def _start_trial(self, trial: Trial) -> bool:
...
+        cached_resources = allocated_resources.resource_request
+
+        staged_resource_count = self._count_staged_resources()
+        if (
+            len(self._cached_resources_to_actor[cached_resources])
+            >= staged_resource_count[cached_resources]
+            # If we don't have any cached actors, cache. We will reconcile
+            # in the step end anyway. This is to avoid non-working actor re-use
+            # if we only generate new trials one by one.
+            and any(v for v in self._cached_resources_to_actor.values())
```
Comment (xwjiang2010) [inline python/ray/tune/execution/ray_trial_executor.py:594]: Sort of tied to the comment at "cleanup_cached_actor", should we allow for one such cached actor per resource type? The logic is to only allow for one across all resource types.

File: python/ray/tune/execution/ray_trial_executor.py [hunk 16]
```diff
@@ -310,43 +316,79 @@ def _stage_and_update_status(self, trials: Iterable[Trial]):
...
+
+            # If the resources are available from the resource manager, return
+            if self._resource_manager.has_resources_ready(
+                resource_request=resource_request
+            ):
                 return trial
 
         return None
 
+    def _maybe_use_cached_actor(self, trial, logger_creator) -> Optional:
```
Comment (xwjiang2010) [inline python/ray/tune/execution/ray_trial_executor.py:359]: When a cached actor is reused, do we need to pop it from newly_added_xxx, in case it's there?

File: python/ray/tune/execution/ray_trial_executor.py [hunk 17]
```diff
@@ -497,12 +533,90 @@ def _start_trial(self, trial: Trial) -> bool:
...
+            "this trial was lost. This is an error in Ray Tune's execution "
+            "logic. Please raise a GitHub issue at "
+            "https://github.com/ray-project/ray/issues"
+        )
+
+    def _maybe_cache_trial_actor(self, trial: Trial) -> bool:
+        if not self._reuse_actors:
+            return False
+
+        if (
```
Comment (xwjiang2010) [inline python/ray/tune/execution/ray_trial_executor.py:577]: For this equation, should we also consider the number of requests that have been filed to ResourceManager?

File: python/ray/tune/execution/ray_trial_executor.py (hunk 16 above)
Comment (krfricke) [reply python/ray/tune/execution/ray_trial_executor.py:359]: As discussed offline, newly added actors are removed from the set in the same iteration they are added

File: python/ray/tune/execution/ray_trial_executor.py (hunk 17 above)
Comment (krfricke) [reply python/ray/tune/execution/ray_trial_executor.py:577]: As discussed offline, it's the callers responsibility to keep track, so we should contain the logic there

File: python/ray/tune/execution/ray_trial_executor.py (hunk 15 above)
Comment (krfricke) [reply python/ray/tune/execution/ray_trial_executor.py]: I've cleaned up the comments and I think it should be clear now

File: python/ray/tune/execution/ray_trial_executor.py (hunk 11 above)
Comment (krfricke) [reply python/ray/tune/execution/ray_trial_executor.py:556]: We are using the `staged_trials` variable in other places (to iterate), so for simplicity lets keep it here. It will likely be removed and definitely refactored in the actor manager execution refactor.

I've added much more extensive documentation for the invariants now - I hope this makes sense! ptal

File: python/ray/tune/execution/ray_trial_executor.py (hunk 12 above)
Comment (krfricke) [reply python/ray/tune/execution/ray_trial_executor.py:319]: See above, we use staged_trials in other locations, and this way we avoid linearly scanning through all resource requirements on every check. This will likely be refactored soon so let's keep it for now

File: python/ray/tune/execution/ray_trial_executor.py (hunk 14 above)
Comment (krfricke) [reply python/ray/tune/execution/ray_trial_executor.py:904]: See above and offline discussion :-)

File: python/ray/tune/execution/ray_trial_executor.py [hunk 18]
```diff
@@ -230,9 +230,6 @@ def __init__(
         self._resources_to_cached_actors: Dict[
             ResourceRequest, List[Tuple[ray.actor.ActorHandle, AcquiredResource]]
         ] = defaultdict(list)
-        self._last_cached_actor_cleanup = float("-inf")
```
Comment (xwjiang2010) [inline python/ray/tune/execution/ray_trial_executor.py]: Nice! Thanks for pushing for reducing internal structures to keep around!

File: python/ray/tune/execution/ray_trial_executor.py [hunk 19]
```diff
@@ -571,12 +568,14 @@ def _maybe_cache_trial_actor(self, trial: Trial) -> bool:
...
-        This is to allow the trial runner to create a trial that needs the resources,
-        (e.g. if we don't allow any pending trials), which will be fulfilled
-        immediately. If no such trial is created, the cached actor is cleaned up
-        in the next trial runner iteration.
+        One exception is the case when we have no cached actors, yet. In that case,
+        we will always cache the actor in this method.
+
+        Later, in `_cleanup_cached_actors`, we will check again if we need this cached
+        actor. That method will keep the actor if we don't have any staged trials,
+        because we don't know at that point if the next trial might require the same
```
Comment (xwjiang2010) [inline python/ray/tune/execution/ray_trial_executor.py:576]: This makes a lot more sense now.

**Comment by xwjiang2010** [comment]
Thanks for being diligent on documentation. I'm looking forward to seeing the complexity of actor reusing further hidden away inside of the ActorManager class for better modularity and testability.

**Review by xwjiang2010 — APPROVED** [review]
